Rework RoomMessageEvent content API again

KitsuneRal opened the pull request:
It's really straightforward now :) `RoomMessageEvent::has<>()` - check that it's possible to extract event content of a given type; `RoomMessageEvent::get<>()` - actually try to get event content of that type. At this point `has()` is not optimised and in some cases effectively creates a content object any way. Can be tackled sometime later.

KitsuneRal committed:
Rework RoomMessageEvent content API again

It's really straightforward now :) RoomMessageEvent::has<>() - check
that it's possible to extract event content of a given type;
RoomMessageEvent::get<>() - actually try to get event content of that
type. At this point has() is not optimised and in some cases effectively
creates a content object any way. Can be tackled sometime later.

nvrWhere commented on Quotient/events/roommessageevent.cpp:
```diff
@@ -23,9 +23,13 @@ constexpr auto RelatesToKey = "m.relates_to"_L1;
+constexpr auto TextTypeId = "m.text"_L1;
+constexpr auto EmoteTypeId = "m.emote"_L1;
+constexpr auto NoticeTypeId = "m.notice"_L1;
+constexpr auto FileTypeId = "m.file"_L1;
+constexpr auto ImageTypeId = "m.file"_L1;
+constexpr auto AudioTypeId = "m.file"_L1;
+constexpr auto VideoTypeId = "m.file"_L1;
```
These are not correct image will be m.image, video mi.video and audio m.audio

KitsuneRal replied:
Oh grrh, copy-pasta.

KitsuneRal committed:
Fix leftovers and copy-pasta around msgtype ids

nvrWhere commented on Quotient/events/roommessageevent.h:
```diff
@@ -62,49 +62,37 @@ class QUOTIENT_API RoomMessageEvent : public RoomEvent {
-    //! \return true, if the message has a data structure corresponding to
-    //!         a file (such as m.file or m.audio); false otherwise
-    bool hasFileContent() const;
+    //! \return A pointer to the object of the requested type if the event has content of this type;
+    //!         nullptr, if the event has no content or the content cannot be cast to this type
+    template <std::derived_from<EventContent::Base> ContentT>
+    std::unique_ptr<ContentT> get() const
+    {
+        return has<ContentT>()
+                   ? std::unique_ptr<ContentT>(static_cast<ContentT*>(content().release()))
+                   : nullptr;
+    }
```
So this leads to a nullptr being returned if I ask for `FileContent` on an Image.

The reason this caught me out is that in NeoChat when getting media info to show in the UI it's done in 2 steps:
1.  Grab all the generic params (by grabbing a `FileContent` currently
2. Cast to specific type to get the unique ones.

Now it wouldn't be hard to change to make this work (get the specialised content object straight away) but it still feels weird that `FileContentBase` doesn't have all the generic params present in each filetype e.g. source, filename

KitsuneRal replied:
Well, yes, `FileContentBase` is effectively a virtual API for `FileInfo` that only happens to expose `url()` at the moment. I wanted to go with as limited set as possible because I still hope to throw away all the code in `eventcontent.*` and replace it with something easier to consume (I have some thoughts around the new hierarchy but they are rather disorganised).

I can probably add `has<FileInfo>()` and `get<FileInfo>()` et al. as well - probably replacing `has/get` for `FileContentBase` as it exposes a subset of `FileInfo`. `FileInfo` is not a polymorphic class though - you cannot call `toJson()` on `std::unique_ptr<FileInfo>`, for one - and I remember a place in Quaternion that uses `FileContentBase` being derived from `Base`. So probably I'll have to keep all three, although it's very confusing.

nvrWhere replied:
As I said I can work round it so unless we add just some simple virtual members to `FileConentBase` it's probably not worth it right now. Just document it clearly that you'll get a null from `get<FileConent>` or other if the file content isn't an exact match e.g. `m.file` for `FileContent`

KitsuneRal replied:
I played around with this. First of all, the very similar concern to yours holds even for the library itself, so fixing this is absolutely warranted. And I think I found a reasonable middle ground - `FileContentBase::commonInfo()` will return a _copy_ of `FileInfo` to handle the NeoChat's case, and it also now has `Thumbnail` so you can show a thumbnail for any file content without even figuring out the type of that content. There are still cases where this is not enough (e.g. updating the file info cannot easily be done; I had to add `RoomMessageEvent::updateFileSource()` to support a pending event case) but I hope we can stop at that for now without affecting the code base further. Commits coming rsn.

KitsuneRal committed:
fileSourceInfoFromJson()

We already have fillJson() for FileSourceInfo, adding a deserialising
function is just logical (and fixes a problem with Thumbnail's EFM
structure never being initialised because EFM had to be supplied as
a separate argument and was forgotten).

KitsuneRal committed:
RoomMessageEvent::updateFileSourceInfo()

Instead of deserialising the FileSourceInfo structure from the event,
updating it and serialising again, just serialise the new structure
and update the content JSON with the result.

KitsuneRal committed:
FileContentBase::commonInfo()

Fixes dealing with common attributes being a problem, described in
https://github.com/quotient-im/libQuotient/pull/812#discussion_r1807418659
